dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,626 @@
+/*
+    DSSCollection contains collection group & completion functionality for DSS.
+    Author: Jeremy Ahrens jer.ahrens@dapperlabs.com
+*/
+
+import NonFungibleToken from "./NonFungibleToken.cdc"
+import MetadataViews from 0xMETADATAVIEWSADDRESS
+
+// The DSSCollection contract
+//
+pub contract DSSCollection: NonFungibleToken {
+
+    // Contract Events
+    //
+    pub event ContractInitialized()
+
+    // NFT Collection Events
+    //
+    pub event Withdraw(id: UInt64, from: Address?)
+    pub event Deposit(id: UInt64, to: Address?)
+
+    // Events
+    //
+    pub event CollectionGroupCreated(
+        id: UInt64,
+        name: String,
+        typeName: String,
+        startTime: UFix64?,
+        endTime: UFix64?,
+        timeBound: Bool
+    )
+    pub event CollectionGroupClosed(id: UInt64)
+    pub event SlotCreated(
+        id: UInt64,
+        collectionGroupID: UInt64,
+        logicalOperator: String,
+        typeName: String,
+        slotType: String
+    )
+    pub event ItemAddedToSlot(
+        itemID: UInt64,
+        itemValue: UInt64,
+        collectionGroupID: UInt64
+    )
+    pub event DSSCollectionNFTMinted(
+        id: UInt64,
+        collectionGroupID: UInt64,
+        serialNumber: UInt64,
+        completedBy: String,
+        completionDate: UFix64,
+        level: UInt64
+    )
+    pub event DSSCollectionNFTBurned(id: UInt64)
+
+
+    // Named Paths
+    //
+    pub let CollectionStoragePath:  StoragePath
+    pub let CollectionPublicPath:   PublicPath
+    pub let AdminStoragePath:       StoragePath
+    pub let MinterPrivatePath:      PrivatePath
+
+    // Entity Counts
+    //
+    pub var totalSupply:                 UInt64
+    pub var nextCollectionGroupID:       UInt64
+    pub var nextSlotID:       UInt64
+
+    // Lists in contract
+    //
+    access(self) let collectionGroupByID: @{UInt64: CollectionGroup}
+    access(self) let slotByID: @{UInt64: Slot}
+
+    // A public struct to access Slot data
+    //
+    pub struct SlotData {
+        pub let id: UInt64
+        pub let collectionGroupID: UInt64
+        pub let logicalOperator: String // (AND / OR)
+        pub let typeName: String // (A.contractAddress.NFT...)
+        pub let slotType: String // (edition.id, edition.tier, play.id)
+        pub var items: {UInt64: AnyStruct}
```

Comment:
I'm lost on why this is a map and not an array.  Maybe it'll explain what I'm missing below.

With the current contract, slotType should be tied to the item.  Otherwise AND becomes useless.  EG:

logicalOperator = AND
slotType = editionId 
items = {1: 1, 2: 100}

You can never have edition id 1 and 100 on the same slot.

You might want

edition.metadata.tier = "Rare" OR edition.metadata.player = "Ryan"

But I don't think you can express that right now.


Is there anything in here for the level?

Should it be like SlotItem as a struct that has a slotType, slotValue, and say level?

Reply:
Ah. the key of the map is edition.id. The value is points. I can switch to an array though because that will make bulk additions easier. I think Deewai had that comment.

{1: 10, 100: 10}

Reply:
ah ok, I still think that there's an issue that you can't say that a player is rare OR ryan with the current layout.

Reply:
Have addressed this with my latest push. Slots are now an array of items.